### PR TITLE
Try to avoid git cloning

### LINKED ISSUE #59: Use `git ls-remote` to get tags instead of full clone
If we want to switch to ex: GitHub Actions (https://github.com/cdnjs/bot-ansible/issues/11), cloning every repository on every run isn't feasible.

We should get the tags with [`git ls-remote`](https://git-scm.com/docs/git-ls-remote.html) and only clone if a new version is available (and we should use `--depth 1`).

### PR DESCRIPTION
**Pull request by klausenbusk**
~~A commit message hasn't been written, the code hasn't been tested and I'm wondering if it can be done with less code duplication~~?

Fix #59

### CODE REVIEW

File: git/git.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ func GetVersions(ctx context.Context, pckg *packages.Package, packageGitcache st
 		gitVersions = append(gitVersions, Version{
 			Tag:       tag,
 			Version:   strings.TrimPrefix(tag, "v"),
-			TimeStamp: packages.GitTimeStamp(ctx, packageGitcache, tag),
+			TimeStamp: packages.GitTimeStamp(ctx, gitpath, tag),
```
Comment (klausenbusk) [inline git/git.go:47]: `GitTimeStamp` will fail when giving a remote URL and "spam" the log. Can we handle that in a nicer way?

Comment (klausenbusk) [reply git/git.go:47]: + calling `GetMostRecentVersion` is just waste of time, as `TimeStamp` is `time.Unix(0, 0)`

**Comment by klausenbusk** [comment]
`git ls-remote --tags git://github.com/ceolter/ag-grid.git` list a lot of tags, but only a few of the newest tags is stored in the `cdnjs` repository. Hmm, can we handle that in a smart way?

**Comment by MattIPv4** [comment]
The bot currently only ever considers the 10 most recent versions from NPM or Git. Honestly, I'd like to be able to get to a state where we're performant enough that we could consider all versions, but for example only the 10 most recent missing per run (so we don't choke the bot).

**Comment by xtuc** [comment]
Currently the bot processes the packages sequencially which makes it slow. It's better to limit to 10 so we don't pull in too many outdated versions, however future versions aren't restricted to 10

**Comment by klausenbusk** [comment]
I just changed the code so we do a [partial clone](https://git-scm.com/docs/partial-clone) (`checker` is currently broken), so the blobs is downloaded on demand. It helps a lot.

**Comment by xtuc** [comment]
> checker is currently broken

What do you mean? can you please send a link?

**Comment by klausenbusk** [comment]
> What do you mean? can you please send a link?

`checker` doesn't currently work, as the "partial clone" logic broke it. Sorry for the confusion.

**Comment by klausenbusk** [comment]
> Currently the bot processes the packages sequencially which makes it slow. It's better to limit to 10 so we don't pull in too many outdated versions, however future versions aren't restricted to 10

Would it be okay to add all the "outdated versions" to the `package.json`? That should reduce the amount of clones greatly.

File: cmd/autoupdate/main.go [hunk 2]
```diff
@@ -199,15 +222,15 @@ func commitNewVersions(ctx context.Context, newVersionsToCommit []newVersionToCo
...
+	updateVersionInCdnjs(ctx, pckg, latestVersion, skippedVersions, packageJSONPath)
 
 	// Add to git the updated package.json
-	packages.GitAdd(ctx, cdnjsPath, path.Join(pckg.Path(), "package.json"))
+//	packages.GitAdd(ctx, cdnjsPath, path.Join(pckg.Path(), "package.json"))
 
-	commitMsg := fmt.Sprintf("Add %s package.json (v%s)", pckg.Name, latestVersion)
-	packages.GitCommit(ctx, cdnjsPath, commitMsg)
+//	commitMsg := fmt.Sprintf("Add %s package.json (v%s)", pckg.Name, latestVersion)
+//	packages.GitCommit(ctx, cdnjsPath, commitMsg)
```
Comment (klausenbusk) [inline cmd/autoupdate/main.go:235]: Ignore this..

File: cmd/autoupdate/main.go [hunk 3]
```diff
@@ -81,13 +95,21 @@ func main() {
...
+		commitPackageVersion(ctx, pckg, latestVersion, skippedVersions, f)
+		if !noUpdate {
+			if len(newVersionsToCommit) > 0 {
+				commitNewVersions(ctx, newVersionsToCommit)
+				commitPackageVersion(ctx, pckg, latestVersion, skippedVersions, f)
+			}
+			fmt.Println(skippedVersions)
+			if (len(skippedVersions) > 0) {
+				commitPackageVersion(ctx, pckg, latestVersion, skippedVersions, f)
+			}
```
Comment (klausenbusk) [inline cmd/autoupdate/main.go:112]: We need a new code path for updating `skippedVersions`.

**Comment by klausenbusk** [comment]
Tested with `getPackages()` hard-coded to return `1/16pixels.json` (all versions is in `cdnjs`):
```
2020/07/05 21:51:48 1/16pixels.json: running git update
2020/07/05 21:51:48 1/16pixels.json: run /usr/bin/git ls-tree --name-only origin/master /home/foobar/Hentet/cdnjs/cdnjs/ajax/libs/16pixels/0.1.2 /home/foobar/Hentet/cdnjs/cdnjs/ajax/libs/16pixels/0.1.3 /home/foobar/Hentet/cdnjs/cdnjs/ajax/libs/16pixels/0.1.4 /home/foobar/Hentet/cdnjs/cdnjs/ajax/libs/16pixels/0.1.5 /home/foobar/Hentet/cdnjs/cdnjs/ajax/libs/16pixels/0.1.6 /home/foobar/Hentet/cdnjs/cdnjs/ajax/libs/16pixels/0.1.7 /home/foobar/Hentet/cdnjs/cdnjs/ajax/libs/16pixels/0.1.8 /home/foobar/Hentet/cdnjs/cdnjs/ajax/libs/16pixels/0.1.9 from /home/foobar/Hentet/cdnjs/cdnjs/ajax/libs/16pixels
2020/07/05 21:51:48 1/16pixels.json: found 8 staged versions
2020/07/05 21:51:48 1/16pixels.json: run /usr/bin/git ls-remote --tags --refs --quiet git://github.com/englishextra/16pixels.git
2020/07/05 21:51:50 1/16pixels.json: found tags in git: [0.1.4 0.1.5 v0.1.2 v0.1.3 v0.1.4 v0.1.5 v0.1.6 v0.1.7 v0.1.8 v0.1.9]
```

Anyone want to take a look? @tc80 (you did comment on the issue)

**Comment by xtuc** [comment]
@klausenbusk i'll go ahead and close this PR, I hope you don't mind but we are focusing on removing git rather than trying to optimize it

**Comment by klausenbusk** [comment]
The purpose of the PR is to optimize the Git autoupdate logic, not the 
cdnjs repo stuff. Are you planing on removing that? ;)

**Comment by tc80** [comment]
@klausenbusk Yeah you're right. I really want to optimize the autoupdate logic/clean up the repo too! 

Let's investigate this after we do our big migration to KV.

Thanks again for all your support and paying so much attention to detail!! 😄

**Comment by tc80** [comment]
@klausenbusk we are now ready to optimize git!! sorry it's been a while 😄

**Comment by klausenbusk** [comment]
> @klausenbusk we are now ready to optimize git!! sorry it's been a while smile

I just did a rebase, I haven't tested the rebased version though.

**Comment by xtuc** [comment]
Since we changed the architecture, git has been optimized

**Pull request closed without merging**